### PR TITLE
Use self-hosted agent

### PR DESCRIPTION
**Pull request by leofang**
While waiting for the CI let me lay out my thoughts and plan... 

The upstream recipe is maintained by John Kirkham and me. But the most challenging part is that we do not have access to GPU resources for testing the recipe, so we've hit hard by some bugs along the way. Since now we have our own clone of the CF build chain, and
1. our ptychography software needs CuPy,
2. we are exploring the possibility of using local machines to build some packages,

it makes perfect sense to use this feedstock as a testbed/playground.

Moreover, as a CuPy developer I would love to build development versions for local benchmarking purposes. This is something CF's policy does not allow but I would love to see happening. If everything works out fine and resource permits, I wanna build
- ~~v7.3.0 (this PR)~~
- v7.4.0 **(UPDATE: this PR)**
- any future official v7 releases
- v8.0.0a1 
- v8.0.0b1
- v8.0.0b2 (just released)
- any development and official releases of v8

However there're some technical issues I'll have to address first.

### CODE REVIEW

**Comment by leofang** [comment]
790b71c is all we need to switch to a local agent!

**Comment by leofang** [comment]
It builds but the test says GPU not found...Looking into this locally...

**Comment by leofang** [comment]
CUDA driver is not loaded in the docker image. It seems we can't use CF's toolchain and instead need to define our own Azure workflow as in NSLS-II/ptycho_gui#90...?

File: .scripts/run_docker_build.sh [hunk 1]
```diff
@@ -59,7 +59,7 @@ if [ -z "${CI}" ]; then
 fi
 
 export UPLOAD_PACKAGES="${UPLOAD_PACKAGES:-True}"
-docker run ${DOCKER_RUN_ARGS} \
+nvidia-docker run ${DOCKER_RUN_ARGS} \
```
Comment (leofang) [inline .scripts/run_docker_build.sh]: This is a machine-specific change that makes the GPUs recognized by docker. `nvidia-docker` is discontinued as NVIDIA drivers are now natively supported by docker, but on our machine somehow it has some problems which I haven't diagnosed, and I noticed the old `nvidia-docker` still works.

**Review by leofang — COMMENTED** [review]
Nice! Test suites are running on GPU! 🎉

File: .scripts/run_docker_build.sh (hunk 1 above)
Comment (leofang) [reply .scripts/run_docker_build.sh]: As docker is heavily used on this machine, I prefer not to fix/update docker for now.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -64,6 +68,10 @@ test:
   requires:
     - pytest
     - mock
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ compiler("cuda") }}  # tests need nvcc
+
```
Comment (leofang) [inline recipe/meta.yaml:78]:
```suggestion
    - scipy

```

File: recipe/run_test.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+except ImportError as e:
+    print('Got ImportError: \n%s' % str(e))
+    print("No GPU available. Exiting without running CuPy's tests.")
+    sys.exit(0)
+
+# Run CuPy's test suite
+import py
+py.test.cmdline.main(["tests/cupy_tests"])
+py.test.cmdline.main(["tests/cupyx_tests"])
+py.test.cmdline.main(["tests/example_tests"])
```
Comment (leofang) [inline recipe/run_test.py]:
```suggestion

```

**Comment by leofang** [comment]
Manually editing `.azure-pipelines/azure-pipelines-linux.yml` can be avoided with conda-forge/conda-smithy#1296!

**Comment by jakirkham** [comment]
Thanks for working on this Leo! If we are able to get this to run using a self-hosted agent, what do you think about adding this to the conda-forge feedstock?

**Comment by leofang** [comment]
Someone has to pay for the GPU resources 😉 In our case this is 1. for my own curiosity and 2. exploring and prepping for future internal needs, possibly involving closed-source and/or patent packages. (Btw I am not sure whether this feedstock should survive, will need to consult @mrakitin after his return. Also, we will migrate to other machines after lockdown is lifted.)

I am not the person who can make the final call (nor would I like to), but my understanding is that some regulations would prohibit our (federal) machines from being exposed to somewhere we don't have sole control/ownership (e.g. CF), so at least it's not possible with this machine. I will be more than happy to set the agents up for CF if access to some resources is granted to me. 

Is it possible with gpuCI? I've been wondering about this for a while...

**Comment by leofang** [comment]
https://github.com/conda-forge/conda-smithy/pull/1296 is merged 🎉

**Comment by leofang** [comment]
The cuTENSOR support is not built due to https://github.com/conda-forge/cupy-feedstock/issues/57.

**Comment by mrakitin** [comment]
> [conda-forge/conda-smithy#1296](https://github.com/conda-forge/conda-smithy/pull/1296) is merged 🎉

Yay! Congrats :)

**Comment by leofang** [comment]
Tests are passed. Let me self-merge and decide what to do next.

**Comment by mrakitin** [comment]
Sounds good to me.

**Pull request merged**